File: Algorithmic/GeometryEngine2D/geometry.go
package main

import (
	"math"
	"sort"
)

// Point represents a 2D point.
type Point struct {
	X, Y float64
}

// CrossProduct returns the cross product of vectors OA and OB.
// Positive if O->A->B is counter-clockwise.
// Negative if clockwise.
// Zero if collinear.
func CrossProduct(o, a, b Point) float64 {
	return (a.X-o.X)*(b.Y-o.Y) - (a.Y-o.Y)*(b.X-o.X)
}

// ConvexHull computes the convex hull of a set of points using Monotone Chain algorithm.
// Returns points in counter-clockwise order.
func ConvexHull(points []Point) []Point {
	n := len(points)
	if n <= 2 {
		return points
	}

	// Sort by X, then Y
	sort.Slice(points, func(i, j int) bool {
		if points[i].X == points[j].X {
			return points[i].Y < points[j].Y
		}
		return points[i].X < points[j].X
	})

	// Build lower hull
	lower := []Point{}
	for _, p := range points {
		for len(lower) >= 2 && CrossProduct(lower[len(lower)-2], lower[len(lower)-1], p) <= 0 {
			lower = lower[:len(lower)-1]
		}
		lower = append(lower, p)
	}

	// Build upper hull
	upper := []Point{}
	for i := n - 1; i >= 0; i-- {
		p := points[i]
		for len(upper) >= 2 && CrossProduct(upper[len(upper)-2], upper[len(upper)-1], p) <= 0 {
			upper = upper[:len(upper)-1]
		}
		upper = append(upper, p)
	}

	// Concatenate (remove duplicate start/end points)
	return append(lower[:len(lower)-1], upper[:len(upper)-1]...)
}

// IsPointInPolygon checks if point p is inside the polygon using Ray Casting algorithm.
// Polygon vertices should be ordered (CW or CCW).
func IsPointInPolygon(p Point, polygon []Point) bool {
	n := len(polygon)
	inside := false
	j := n - 1
	for i := 0; i < n; i++ {
		// Check if ray crosses edge (polygon[i], polygon[j])
		if ((polygon[i].Y > p.Y) != (polygon[j].Y > p.Y)) &&
			(p.X < (polygon[j].X-polygon[i].X)*(p.Y-polygon[i].Y)/(polygon[j].Y-polygon[i].Y)+polygon[i].X) {
			inside = !inside
		}
		j = i
	}
	return inside
}

// Area calculates the polygon area using the shoelace formula.
func Area(polygon []Point) float64 {
	area := 0.0
	n := len(polygon)
	j := n - 1
	for i := 0; i < n; i++ {
		area += (polygon[j].X + polygon[i].X) * (polygon[j].Y - polygon[i].Y)
		j = i
	}
	return math.Abs(area / 2.0)
}

File: Algorithmic/GeometryEngine2D/python/geometry.py
"""Geometry Engine 2D.

Convex Hull (Monotone Chain), Point in Polygon (Ray Casting).
"""

from typing import List, Tuple

Point = Tuple[float, float]

def cross_product(o: Point, a: Point, b: Point) -> float:
    return (a[0] - o[0]) * (b[1] - o[1]) - (a[1] - o[1]) * (b[0] - o[0])

def convex_hull(points: List[Point]) -> List[Point]:
    """Compute Convex Hull using Monotone Chain algorithm. O(N log N)."""
    n = len(points)
    if n <= 2:
        return points

    points = sorted(points)

    upper = []
    for p in points:
        while len(upper) >= 2 and cross_product(upper[-2], upper[-1], p) <= 0:
            upper.pop()
        upper.append(p)

    lower = []
    for p in reversed(points):
        while len(lower) >= 2 and cross_product(lower[-2], lower[-1], p) <= 0:
            lower.pop()
        lower.append(p)

    return upper[:-1] + lower[:-1]

def point_in_polygon(poly: List[Point], p: Point) -> bool:
    """Check if point p is inside polygon poly (Ray Casting)."""
    x, y = p
    n = len(poly)
    inside = False
    p1x, p1y = poly[0]
    for i in range(n + 1):
        p2x, p2y = poly[i % n]
        if y > min(p1y, p2y):
            if y <= max(p1y, p2y):
                if x <= max(p1x, p2x):
                    if p1y != p2y:
                        xinters = (y - p1y) * (p2x - p1x) / (p2y - p1y) + p1x
                    if p1x == p2x or x <= xinters:
                        inside = not inside
        p1x, p1y = p2x, p2y
    return inside
